### ISSUE TITLE
TLauncher Account Support (mostly to be able to see skins from TLauncher players)

### ISSUE THREAD

**Issue by NiDzA-12** [AfoninZ/MultiMC5-Cracked#67]
### Role

In multiplayer

### Suggestion

To be able to see skins from players that use TLauncher

### Benefit

It would allow for players that use a "default" cracked launcher (TLauncher) and people that prefer th multimc gui (people that use your launcher)

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

Thanks in Advance

**Comment by AndreVuillemot160** [comment]
This is pretty shitty idea IMO
Especially, since these guys don't allow you delete your own account from their system.

**Comment by AfoninZ** [comment]
Wouldn't call this shitty on sight, but unfortunately it's not the top priority task right now. I've got to update the launcher to the newest MMC codebase and then work on newer features.

**Comment by AndreVuillemot160** [comment]
I think it is shitty due the developers behind TLauncher devs are pretty shady. The thing is, is that there are two versions of TLauncher. With account one is a fake, and meanwhile there's original one developed still but turikhay aka original dev.

**Comment by AndreVuillemot160** [comment]
From dev side, these guys didn't wrote any documentation about how to implement this.

**Comment by NiDzA-12** [comment]
Yeah it is mostly because my friends that use TLauncher don't see my skin
but personally i don't care about their perspective so thanks for the
feedback

**Comment by AndreVuillemot160** [comment]
https://vk.com/tl_mc_launcher here is also their original vk group

**Comment by lebestnoob** [comment]
While this comment isn't necessarily contributing to the project, I found a forge and fabric mod that allows you to view TLauncher skins. It could be used as a temporary solution to the problem while this feature, if it even is possible at all, gets added. The mod is called [CustomSkinLoader](https://github.com/xfl03/MCCustomSkinLoader). I tried it out on some cracked servers and it displayed their skins. I even doubled checked the skin that was shown to me by inputting their username on TLauncher's skins catalog and they matched.

[Download link](https://www.curseforge.com/minecraft/mc-mods/customskinloader/files) 

One thing you may want to do is modify the mod's config file, which is located in .minecraft/CustomSkinLoader/CustomSkinLoader.json (Click Minecraft folder button on an instance in the launcher), so it will only display TLauncher skins. The loadlist in the json should look like this: 
```
  "loadlist": [
    {
      "name": "TLauncher",
      "type": "ElyByAPI",
      "root": "https://auth.tlauncher.org/skin/profile/texture/login/"
    }
  ],
```

**Comment by zaphod77** [comment]
That's quite interesting.  

It TLauncher really uses elyby api, then it's as simple as adding their auth servers to support?

and the mod seems to understand HD skins too?  but i see notign in te tlauncher section about capes.

This is actually quite useful. it's got source too.

INTERESTING...

**Comment by AndreVuillemot160** [comment]
I guess they self-host ely.by lol

**Comment by Samisafool** [comment]
Keep in mind, DO NOT INSTALL TLAUNCHER. The tlaun.ch one is fine, but the one that pops up on google is spyware. Use it in a VM if you must.

**Comment by AndreVuillemot160** [comment]
You can also read the story here https://minecraftmain.ru/2018/02/tlauncher-story (on russian so use translator to english like yandex translate)

**Comment by Samisafool** [comment]
> they do give you instructions on how to patch tlauncher skins

where?

**Comment by zaphod77** [comment]
They have you edit a manifest file and then it will show the TL icon in the launcher for the installation, and the skins will work.  also it used to be implemented through some sort of mod that can still be found on curseforge.

**Comment by Samisafool** [comment]
> They have you edit a manifest file and then it will show the TL icon in the launcher for the installation, and the skins will work. also it used to be implemented through some sort of mod that can still be found on curseforge.

Then it's only on tlauncher. I found the mod, [here](https://www.curseforge.com/minecraft/mc-mods/tl-skin-mod), but it's closed source. Anyone know where the auth server is?

**Comment by zaphod77** [comment]
Hopefully adding in spyware-free tlauncher.org support won't piss them off too much.  :)  There's a couple other alternate skinning apis, but i don't know if they do auth as well.

**Comment by Samisafool** [comment]
If they do, ask them to cope

**Comment by zaphod77** [comment]
it won't bypass their premium, just their launcher ads.

which if they are aware of they might object to.

Not my problem, though. :)

**Comment by Samisafool** [comment]
What're they gonna do about it? Cry to the pirate government or whatever? lol

**Comment by zaphod77** [comment]
anyone figure this out? I'm not exactly up for testing it, but it would make some peopel happy.

**Comment by Samisafool** [comment]
If I can figure out their auth URL, probably possible.

**Comment by zaphod77** [comment]
GOT iT.

skin.authentication.authenticate=https://auth.tlauncher.org/skin/authentication/authenticate
skin.authentication.refresh=https://auth.tlauncher.org/skin/authentication/refresh

**Comment by Samisafool** [comment]
soooo... not self-hosted elyby. sad

**Comment by zaphod77** [comment]
it's Yggdrasil at https://auth.tlauncher.org/skin/authentication/

complete with /authenticate and /refresh endpoints. so should be no trouble to implement.

**Comment by Samisafool** [comment]
ohh, I thought it was like ely.by

**Comment by maximmasterr** [comment]
We use yggdrasil emulation in ely.by, so tlauncher authentic wil be easy to implement

**Comment by Samisafool** [comment]
When I try to do it, I get `Authentication server attempted to change the client token. This isn't supported.`

**Comment by zaphod77** [comment]
that's what i thoguht. from what i gather most alternative launchers use yggdrasil as well since it's an open auth api. you just need to extract their jar file with winzip and get their auth URLs, get their skin url, identify their actual skin api, and poof. 

That mod supports a few more skin APIs.

You can also run suspicious launchers sandboxed to fetch their libraries so you can decompile them, if a text scan isn't enough to get the auth and skin urls.

**Comment by Samisafool** [comment]
>  This seems like an intentional change to try and stop people from interoperating with them.

So how hard will it be to get it to work?

**Comment by zaphod77** [comment]
as i said, i think you just have to change the client to not provide a token that hasn't been sent from the server yet, and to track it separately for each account. If i had a working build environment on my computer i'd try myself.

**Comment by Samisafool** [comment]
Any luck here?

**Comment by zaphod77** [comment]
did anyone who actually has a tlauncher.org account figure this out?

**Comment by KindaNeatdoe** [comment]
Any updates?

**Comment by pokcreator** [comment]
Hi, maybe someone knows where TL checks the user's UUID

**Comment by Neptune650** [comment]
I'll see the possibility of implementing this by reverse engineering TLauncher, nothing confirmed by now though, needless to say we already have ely.by support so the one tlaun.ch uses should work just fine right now.

**Comment by Neptune650** [comment]
In order to properly implement this feature it's important to know from which URL TLauncher gets the profile (UUID) of an account, any idea?
